cli/ops/lint.rs

```diff
@@ -3,11 +3,163 @@
 use deno_ast::MediaType;
 use deno_ast::ModuleSpecifier;
 use deno_ast::ParseDiagnostic;
+use deno_ast::SourceRange;
+use deno_ast::SourceTextInfo;
+use deno_ast::SourceTextProvider;
 use deno_core::op2;
+use deno_core::OpState;
+use deno_lint::diagnostic::LintDiagnostic;
+use deno_lint::diagnostic::LintDiagnosticDetails;
+use deno_lint::diagnostic::LintDiagnosticRange;
+use deno_lint::diagnostic::LintFix;
+use deno_lint::diagnostic::LintFixChange;
+use tokio_util::sync::CancellationToken;
 
 use crate::tools::lint;
+use crate::tools::lint::PluginLogger;
 
-deno_core::extension!(deno_lint, ops = [op_lint_create_serialized_ast,],);
+deno_core::extension!(
+  deno_lint_ext,
+  ops = [
+    op_lint_create_serialized_ast,
+    op_lint_report,
+    op_lint_get_source,
+    op_is_cancelled
+  ],
+  options = {
+    logger: PluginLogger,
+  },
+  // TODO(bartlomieju): this should only be done,
+  // if not in the "test worker".
+  middleware = |op| match op.name {
+    "op_print" => op_print(),
+    _ => op,
+  },
+  state = |state, options| {
+    state.put(options.logger);
+    state.put(LintPluginContainer::default());
+  },
+);
+
+deno_core::extension!(
+  deno_lint_ext_for_test,
+  ops = [op_lint_create_serialized_ast, op_is_cancelled],
+  state = |state| {
+    state.put(LintPluginContainer::default());
+  },
+);
+
+#[derive(Default)]
+pub struct LintPluginContainer {
+  pub diagnostics: Vec<LintDiagnostic>,
+  pub source_text_info: Option<SourceTextInfo>,
+  pub specifier: Option<ModuleSpecifier>,
+  pub token: CancellationToken,
+}
+
+impl LintPluginContainer {
+  pub fn set_cancellation_token(
+    &mut self,
+    maybe_token: Option<CancellationToken>,
+  ) {
+    let token = maybe_token.unwrap_or_default();
+    self.token = token;
+  }
+
+  pub fn set_info_for_file(
+    &mut self,
+    specifier: ModuleSpecifier,
+    source_text_info: SourceTextInfo,
+  ) {
+    self.specifier = Some(specifier);
+    self.source_text_info = Some(source_text_info);
+  }
+
+  fn report(
+    &mut self,
+    id: String,
+    message: String,
+    hint: Option<String>,
+    start: usize,
+    end: usize,
+    fix: Option<LintReportFix>,
+  ) -> Result<(), LintReportError> {
+    let source_text_info = self.source_text_info.as_ref().unwrap();
+    let specifier = self.specifier.clone().unwrap();
+    let source_range = source_text_info.range();
+    let start_pos = source_text_info.start_pos();
+    let diagnostic_range = SourceRange::new(start_pos + start, start_pos + end);
+    if !source_range.contains(&diagnostic_range) {
+      return Err(LintReportError::DiagnosticIncorrectRange {
+        // JS uses 1-based indexes
```

Comment:
Can we change this to 0-based?

Reply:
AFAIU we wanted to use 1-based to mimic ESLint API that also uses 1-based offsets

Reply:
Maybe we should just do zero if we're not going for backwards compatibility?

Reply:
Looks like this is what eslint does: https://eslint.org/docs/latest/extend/custom-rules#reporting-problems

Reply:
Ah, I see this is just for the error shown to the user. I misunderstood. Probalby we should show line and column numbers here?

Reply:
How can I show line and column number here if the passed range is not within the `SourceRange`?